libs/libdsp/lib_avg.c

```diff
@@ -0,0 +1,92 @@
+/****************************************************************************
+ * libs/libdsp/lib_avg.c
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.  The
+ * ASF licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <assert.h>
+#include <dsp.h>
+#include <string.h>
+
+/* Based on video explanation of Dr. Shane Ross:
+ * https://www.youtube.com/watch?v=HCd-leV8OkU
+ */
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: avg_filter_data_init
+ *
+ * Description:
+ *   Initialize the data struct used to store prev_avg and k parameter.
+ *
+ * Input Parameters:
+ *   data     - pointer to avg_filter_data_s
+ *   prev_avg - initial value of prev_avg
+ *   k        - initial value of k counter
+ *
+ * Returned Value:
+ *   None
+ *
+ ****************************************************************************/
+
+void avg_filter_data_init(FAR struct avg_filter_data_s *data,
+                          float prev_avg, float k)
+{
+  data->prev_avg = prev_avg;
+  data->k        = k;
+}
+
+/****************************************************************************
+ * Name: avg_filter
+ *
+ * Description:
+ *   Calculate the recurring average of a signal in the time
+ *
+ * Input Parameters:
+ *   prev_avg - pointer to previous average variable
+ *   k - pointer to k counter variable
+ *   x - current signal value
+ *
+ * Returned Value:
+ *   Average value
+ *
+ ****************************************************************************/
+
+float avg_filter(FAR struct avg_filter_data_s *data, float x)
+{
+  float alpha;
+  float avg;
+
+  LIBDSP_DEBUGASSERT(data != NULL);
+
+  alpha = (data->k - 1.0f) / data->k;
```

Comment:
Potential division by zero. Maybe can be changed to
```suggestion
  alpha = data->k / (data->k + 1.0f);
```
?

Reply:
It is not the same: imagine k = 1, in the original code alpha should be 0. In your suggestion alpha will be 0.5f

Reply:
With current code it will be either if k is 0 we will have either crash or undefined behaviour. In your example it depends how do we count... Same as array indexing from 0 or from 1 :)
In my proposal the k == 0 is the case of k == 1 in your case. I mean that counter is shifted by 1.0f

Reply:
Understood your point, but k is the number of the current element, not the index. I'm not expert in DSP and tried to follow the math examples, maybe @raiden00pl could make some suggestions here because he has more experience than me.

Reply:
At the very least, maybe ask `if (data->k)` before dividing; however, this will compare to exactly zero. Maybe we need a "smallest allowed" value and compare against that, using fabsf to handle positive and negative values. What happens if `data->k` is too close to zero? Not sure, but perhaps assigning a value that is determined "safe" (even if mathematically incorrect) is better than crashing.

Reply:
@acassis `k` is only incremented by 1 and there is no problem with overflow for float (absorption property of float: `FLT_MAX + 1 = FLT_MAX)`, so the only way to get div by 0 is for the user to provide invalid initial data. 
Then a simple check in the initialization function is enough `LIBDSP_DEBUGASSERT(k > 0.0f)`